### PR TITLE
Make the ReLU example clear

### LINKED ISSUE #176: Understanding `ChainRulesCore.rrule` in custom ReLU example
Hi, developers! Again, it's me, Jinrae.

After understanding the ReLU example (#168), I'm trying to "use" DiffOpt.jl.
Unfortunately, I didn't get it properly.

If I understand the mechanism, it might be possible for me to add my understandings to the documentation (as a newbie).

# What I want to do using DiffOpt.jl
I want to make my custom neural network differentiable via Flux.jl (namely `my_NN`), which contains a convex optimisation problem.
For example, the loss function would be the mse error of
```julia
y - my_NN_optimal(x)
```
where `y` is the label, `x` is condition, and `my_NN_optimal(x) = my_NN(x, u_)` where `u_` is the minimised of `u -> my_NN(x, u)`.
To update `my_NN`'s parameters, I have to get some `rrule`.

# My understanding up to now
## What users can do using DiffOpt.jl
First, I've found in [Introduction](https://jump.dev/DiffOpt.jl/dev/intro/) that DiffOpt.jl can be used for differentiating the optimal solution to given parametrised convex optimisation problem with respect to problem parameters.
For example, let `x_` and `p` be the optimal solution and problem parameters, respectively.
It is possible to make an `rrule` for `dx_ / dp` using DiffOpt.jl.

## Strategy
As presented in some pages in docs, for example, [custom ReLU example](https://jump.dev/DiffOpt.jl/dev/examples/custom-relu/), I may have to add some `rrule`s (reverse-mode differentiation), which will be imported from [ChainRulesCore.jl](https://github.com/JuliaDiff/ChainRulesCore.jl).

## Where I felt stuck in
I'm struggling with the following things.
### What do `dx` and `dy` mean in [the custom ReLU example](https://jump.dev/DiffOpt.jl/dev/examples/custom-relu/)?
In the tutorial, especially where `ChainRulesCore.rrule` is defined, I cannot exactly follow what each line means.
It is probably due to my poor understanding of DiffOpt.jl and related packages, e.g., JuMP.

For example, this line means "performing optimisation" (with manually written function `matrix_relu`).
```julia
    pv = matrix_relu(y, model = model)
```
Then, what do the following lines mean? I mean,
- What is `dx`?
- I still don't understand even what `BackwardInVariablePrimal` is for.
```julia
            MOI.set.(
                model,
                DiffOpt.BackwardInVariablePrimal(),
                x,
                dx[:, i]
            ) # set sensitivities
```

Also, the following outputs `obj_exp` (seems an expression) but I don't get any clue about what it means.
```julia
            obj_exp = MOI.get(
                model,
                DiffOpt.BackwardOutObjective()
            ) # return grdiente wrt objective function parameters
```
For example, some outputs from REPL say
```julia
obj_exp = MOI.get(model, DiffOpt.BackwardOutObjective()) = -0.2500000001711336 x[1]² - 0.5 x[1]
-0.2500000001711336 x[1]² - 0.5 x[1]
```
so it is a quadratic expression of `x`. 
- What is `BackwardOutObjective` for?
- Why is it "the gradient wrt objective function parameters"?
- What's objective function parameters in this tutorial?

And some other questions:
- Which scheme do I have to choose among [two schemes](https://jump.dev/DiffOpt.jl/dev/manual/)?
- How can I readily transform my convex optimisation problem into a conic form (QP form would probably be restrictive in my case)?

Sorry for the verbose questions because of my lack of background for differentiable convex programming.
Thanks always.

### PR DESCRIPTION
**Pull request by JinraeKim**
Resolve #176.
Extended from #177.

I tried to make the ReLU example clear.

### CODE REVIEW

**Comment by joaquimg** [comment]
Lets only edit the ReLU one here? we can leave the other tutorial for the other PR

**Comment by JinraeKim** [comment]
> Lets only edit the ReLU one here? we can leave the other tutorial for the other PR

Ok, I'll do it soon.

**Comment by JinraeKim** [comment]
Sorry but force-push seems not to work.
I have to leave now so I'll fix it later.

**Comment by joaquimg** [comment]
just copy paste the original file

**Comment by JinraeKim** [comment]
@joaquimg Here you are.

**Review by joaquimg — APPROVED** [review]

File: docs/src/examples/custom-relu.jl [hunk 1]
```diff
@@ -64,9 +64,9 @@ function ChainRulesCore.rrule(
             obj_exp = MOI.get(
                 model,
                 DiffOpt.BackwardOutObjective()
-            ) # return grdiente wrt objective function parameters
-            dy[:, i] = JuMP.coefficient.(obj_exp, x) # coeff of `x` in -2x'y
-            dy[:, i] = -2 * dy[:, i]
+            ) # return gradient wrt objective function parameters
+            dq[:, i] = JuMP.coefficient.(obj_exp, x) # coeff of `x` in q'x = -2x'y
+            dy[:, i] = -2 * dq[:, i]  # ∵ dq/dy = -2
```
Comment (blegat) [inline docs/src/examples/custom-relu.jl:69]: If `dq/dy = -2` then it should be `dy[:, i] = -dq[:, i] / 2`.

Comment (joaquimg) [reply docs/src/examples/custom-relu.jl:69]: good catch

File: docs/src/examples/custom-relu.jl [hunk 2]
```diff
@@ -34,7 +34,7 @@ function matrix_relu(
         @objective(
             model,
             Min,
-            x'x -2x'y[:, i]
+            x'x -2x'y[:, i]  # x' Q x + q'x with Q = I, q = -2y
```
[the remark this answers is not in the document]
Comment (joaquimg) [reply docs/src/examples/custom-relu.jl:37]: this might be clearer

File: docs/src/examples/custom-relu.jl (hunk 1 above)
Comment (joaquimg) [reply docs/src/examples/custom-relu.jl:69]:
```suggestion
            dy[:, i] = -2dq[:, i]/2  # q = -2y, so dq/dy = -2 and dy = -dq/2
```

**Review by joaquimg — CHANGES_REQUESTED** [review]
Following @blegat catch

File: docs/src/examples/custom-relu.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (JinraeKim) [reply docs/src/examples/custom-relu.jl:37]: I agreee

File: docs/src/examples/custom-relu.jl (hunk 1 above)
Comment (JinraeKim) [reply docs/src/examples/custom-relu.jl:69]: Nooooooo, that's why I was really confused with this package.
As I pointed out in [a comment, #178](https://github.com/jump-dev/DiffOpt.jl/issues/176#issuecomment-989469397), the `dy` actually refers to `dl/dy` for some output `l` (e.g., loss).

That is, the input of the pullback, `dx`, means `dl/dx`.
So `dy` and `dq` correspond to `dl/dy` and `dl/dq`, respectively.
Therefore, to convert `dq` to `dy`, we need this calculation:
`dl/dy = dl/dq * dq/dy`,
which is notationally opposite to the mathematical expression of `dy = -2 * dq` (sooo confusing).

Comment (JinraeKim) [reply docs/src/examples/custom-relu.jl:69]: For example, in thie ReLU example, we will get `dl/dy`, denoted by `dy` in the code.
When letting `dx` (actually means `dl/dx`) be 1, we will get
`dl/dy = dl/dx * dx/dy = dx/dy`.

Since the activation ReLU provides a subgradient of
\in {0}, if `y < 0`
\in [0, 1], if `y = 0`
\in {1}, if `y > 0`
so whether the output is correct can be checked by applying any `y` into as
```julia
    m = 1
    d = 1
    y = ones(m, d)
    @show matrix_relu(y)
    pv, pullback_matrix_relu = _rrule(y)  # corresponds to ChainRulesCore.rrule
    dx = ones(m, d)  # given dl/dx
    pullback_matrix_relu(dx)  # dl/dy = dl/dx * dx/dy = dx/dy
```
gives us
```julia
dy[:, i] = -2 * dq[:, i] = [1.0]
```
which is correct.

Comment (JinraeKim) [reply docs/src/examples/custom-relu.jl:70]: To make it clear, I added it to docs.

File: docs/src/examples/custom-relu.jl [hunk 3]
```diff
@@ -53,6 +53,7 @@ function ChainRulesCore.rrule(
     function pullback_matrix_relu(dx)
         x = model[:x]
         dy = zeros(T, size(dx))
+        dq = zeros(T, size(dx))  # for step-by-step explanation
```
Comment (JinraeKim) [inline docs/src/examples/custom-relu.jl:56]: Added for pre-assignment.

**Review by JinraeKim — COMMENTED** [review]
Please added contents for detailed explanation

File: docs/src/examples/custom-relu.jl (hunk 1 above)
Comment (joaquimg) [reply docs/src/examples/custom-relu.jl:70]: might well be, i'd have to think more carefully.
if dy is actually `dl/dy,` we have to rename it to `dl_dy`.
same for `dq`.

Comment (joaquimg) [reply docs/src/examples/custom-relu.jl:70]: same for `dx`

Comment (JinraeKim) [reply docs/src/examples/custom-relu.jl:70]: It would be much better to use some different symbols (for example, \dot for forward mode and \overline for reverse mode, e.g., \dot{x}) as [ChainRulesCore.jl](https://juliadiff.org/ChainRulesCore.jl/dev/#Reverse-mode-AD-rules-(rrules)) to distinguish them from `dfoo`.

To avoid some Unicode issues, how 'bout `_x` for forward mode and `x_` for reverse mode?
That is, `_x = dx/da` for some `a`, and `x_ = dl/dx` for some `l`.

Comment (joaquimg) [reply docs/src/examples/custom-relu.jl:70]: I dont see why `x_` is better than `dl_dx` to represent `dl/dx`. The first requires that the user reads a manual saying `x_ = dl/dx`, while the second is much close to what is meant.

Comment (JinraeKim) [reply docs/src/examples/custom-relu.jl:70]: You're right, it's clearer. Honestly, both of them are fine to me :)
But in that case, users should know what `l` refers to as well.
So we would have to put some additional comments either which one is taken.

Comment (joaquimg) [reply docs/src/examples/custom-relu.jl:70]: I agree with the need for additional comments in both options.

File: docs/src/examples/custom-relu.jl (hunk 3 above)
Comment (JinraeKim) [inline docs/src/examples/custom-relu.jl:56]:
```suggestion
        dl_dy = zeros(T, size(dl_dx))
        dl_dq = zeros(T, size(dl_dx))  # for step-by-step explanation
```

File: docs/src/examples/custom-relu.jl [hunk 4]
```diff
@@ -53,6 +53,7 @@ function ChainRulesCore.rrule(
     function pullback_matrix_relu(dx)
```
Comment (JinraeKim) [inline docs/src/examples/custom-relu.jl:53]:
```suggestion
    function pullback_matrix_relu(dl_dx)
    # some value from the backpropagation (e.g., loss) is denoted by `l`
```

File: docs/src/examples/custom-relu.jl [hunk 5]
```diff
@@ -64,15 +65,24 @@ function ChainRulesCore.rrule(
             obj_exp = MOI.get(
                 model,
                 DiffOpt.BackwardOutObjective()
-            ) # return grdiente wrt objective function parameters
-            dy[:, i] = JuMP.coefficient.(obj_exp, x) # coeff of `x` in -2x'y
-            dy[:, i] = -2 * dy[:, i]
+            ) # return gradient wrt objective function parameters
+            dq[:, i] = JuMP.coefficient.(obj_exp, x) # coeff of `x` in q'x = -2y'x
+            dy[:, i] = -2 * dq[:, i]  # ∵ dq/dy = -2
```
Comment (JinraeKim) [inline docs/src/examples/custom-relu.jl:70]:
```suggestion
            dl_dq[:, i] = JuMP.coefficient.(obj_exp, x) # coeff of `x` in q'x = -2y'x
            dl_dy[:, i] = -2 * dl_dq[:, i]  # ∵ dq/dy = -2
```

File: docs/src/examples/custom-relu.jl [hunk 6]
```diff
@@ -64,15 +65,24 @@ function ChainRulesCore.rrule(
...
 
+# The line, `dy[:, i] = -2 * dq[:, i]` might be really confusing as
+# ```math
+# q = -2y.
+# ```
+# However, the `dfoo` in pullback refers to `dl/dfoo` for some output `l`, for example, a loss.
+# That is, `dy` actually means `dl/dy`, and therefore `dl/dy = dl/dq * dq/dy = dl/dq * (-2)` is correct
+# For more details, visit [Introduction, ChainRulesCore.jl](https://juliadiff.org/ChainRulesCore.jl/dev/).
+
+
```
Comment (JinraeKim) [inline docs/src/examples/custom-relu.jl:85]:
```suggestion
# For more details about backpropagation, visit [Introduction, ChainRulesCore.jl](https://juliadiff.org/ChainRulesCore.jl/dev/).


```

File: docs/src/examples/custom-relu.jl [hunk 7]
```diff
@@ -64,15 +65,24 @@ function ChainRulesCore.rrule(
...
                 model,
                 DiffOpt.BackwardOutObjective()
-            ) # return grdiente wrt objective function parameters
-            dy[:, i] = JuMP.coefficient.(obj_exp, x) # coeff of `x` in -2x'y
-            dy[:, i] = -2 * dy[:, i]
+            ) # return gradient wrt objective function parameters
+            dq[:, i] = JuMP.coefficient.(obj_exp, x) # coeff of `x` in q'x = -2y'x
+            dy[:, i] = -2 * dq[:, i]  # ∵ dq/dy = -2
         end
         return (ChainRulesCore.NoTangent(), dy,)
```
Comment (JinraeKim) [inline docs/src/examples/custom-relu.jl:72]:
```suggestion
        return (ChainRulesCore.NoTangent(), dl_dy,)
```

**Review by JinraeKim — COMMENTED** [review]
@joaquimg
`dfoo` is replaced by `dl_dfoo` in pullback.

File: docs/src/examples/matrix-inversion-manual.jl [hunk 8]
```diff
@@ -159,7 +159,7 @@ dx = MOI.get.(
     model,
     DiffOpt.ForwardOutVariablePrimal(),
     x,
-)  # ∂x/∂h
+)
 
 using Test                                  #src
 @test dx ≈ [0.25 ,0.75] atol=1e-4 rtol=1e-4 #src
```
Comment (JinraeKim) [inline docs/src/examples/matrix-inversion-manual.jl:165]:
```suggestion
@test dl_dx ≈ [0.25 ,0.75] atol=1e-4 rtol=1e-4 #src
```
Fix [the test failure](https://github.com/jump-dev/DiffOpt.jl/runs/4508768666?check_suite_focus=true)

**Comment by JinraeKim** [comment]
@joaquimg 
I'm confused with why rebase-and-forced-push branch shows so many file changes including #167 ...
Did I messed up?

**Comment by JinraeKim** [comment]
It seems good :)
